pkg/webhook/controlplane/handler.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"net/http"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+)
+
+// newHandler creates a new handler for the given types, using the given validator, and logger.
+func newHandler(mgr manager.Manager, types []runtime.Object, validator Validator, logger logr.Logger) (*handler, error) {
+	// Build a map of the given types keyed by their GVKs
+	typesMap, err := buildTypesMap(mgr, types)
+	if err != nil {
+		return nil, err
+	}
+
+	// Creaate and return a handler
+	return &handler{
+		typesMap:  typesMap,
+		validator: validator,
+		logger:    logger.WithName("handler"),
+	}, nil
+}
+
+type handler struct {
+	typesMap  map[metav1.GroupVersionKind]runtime.Object
+	validator Validator
+	decoder   types.Decoder
+	logger    logr.Logger
+}
+
+// InjectDecoder injects the decoder into the handler.
+func (h *handler) InjectDecoder(d types.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+// Handle handles the given admission request.
+func (h *handler) Handle(ctx context.Context, req types.Request) types.Response {
+	ar := req.AdmissionRequest
+
+	// Decode object
+	obj := h.typesMap[ar.Kind].DeepCopyObject()
```

Comment:
Shouldn't we check whether `ar.Kind` is part of `typesMap` to prevent panics?

Reply:
Good idea, although since the types are used also to configure the webhook server, the panic situation should never happen in practice (assuming there are no bugs in the rest of the code). Added such a check just in case.